File: Data-Structure-I/566-Reshape-the-Matrix.cpp
// solution 1:
class Solution {
public:
    vector<vector<int>> matrixReshape(vector<vector<int>>& mat, int r, int c) {
        if(mat.size()*mat[0].size() != r*c) return mat;
        vector<vector<int>> reshape;
        int index = 0;
        vector<int> row;
        for(int i=0; i<mat.size(); ++i){
            for(int j=0; j<mat[i].size(); ++j){
                row.push_back(mat[i][j]);
                index++;
                
                if(index==c){
                    reshape.push_back(row);
                    index = 0;
                    row.clear();
                }
            }
        }
        return reshape;
    }
};

// solution 2:
class Solution {
public:
    vector<vector<int>> matrixReshape(vector<vector<int>>& mat, int r, int c) {
        if(mat.size()*mat[0].size() != r*c) return mat;
        vector<vector<int>> reshape(r, vector<int>(c));
        int rows = 0, cols = 0;;
        for(int i=0; i<mat.size(); ++i){
            for(int j=0; j<mat[i].size(); ++j){
                if(cols == c){
                    cols = 0;
                    rows++;
                }
                reshape[rows][cols] = mat[i][j];
                cols++;
            }
        }
        return reshape;
    }
};

File: Data-Structure-I/566-Reshape-the-Matrix.py
# solution 1:
class Solution:
    def matrixReshape(self, mat: List[List[int]], r: int, c: int) -> List[List[int]]:
        if len(mat)*len(mat[0]) != r*c:
            return mat
        reshape = []
        row = [] 
        index = 0
        for i in mat:
            for j in i:
                row.append(j)
                index += 1
                if index == c:
                    reshape.append(row)
                    row = []
                    index = 0
        return reshape

# solution 2:
class Solution:
    def matrixReshape(self, mat: List[List[int]], r: int, c: int) -> List[List[int]]:
        if len(mat)*len(mat[0]) != r*c:
            return mat
        reshape = [[None for x in range(c)] for y in range(r)]
        
        i, j = 0, 0
        
        for row in range(r):
            for col in range(c):
                reshape[row][col] = mat[i][j]
                j += 1
                
                if j == len(mat[0]):
                    i += 1
                    j = 0
        return reshape
